README.md
# README for AKA

## Purpose
AKA exists to allow players with multiple characters in World of Warcraft to be more easily recognized by those they communicate with, regardless of which character they are currently playing.


## Quick Start
Here are the steps to have the name of your main or whatever people know you by showing up on all your alts for a given faction on a given realm:

1. Log in to an alt on the desired faction and realm in WoW.
2. Type `/aka guild` into the chat bar and press the Enter key.
3. Type `/aka main <name>` (where name is the actual name of your main without brackets) into the chat bar and press the Enter key.

After these three steps, you should see whatever value you entered for main appear in parentheses after your current character name in guild chat messages for any character in that faction on that realm. If you are logged in as a character with the name matching the value you entered for main, AKA will automatically leave out the parenthetical display of the same name.

NOTE: Technically you could log in to your main and perform steps two and three. However, if your set main to be the actual name of your main (as many do), then testing the appearance of the parenthetical aka will require logging out and then logging over to an alt.


## Usage

The following commands are used in the input field of the chat window in the WoW client.

`/aka` - Displays basic help information

`/aka help` - Displays basic help information

`/aka guild` - Toggle display of main on/off for guild chat for the current faction on the current realm

`/aka whisper` - Toggle display of main on/off for whispers for the current faction on the current realm

`/aka <number>` - Toggle display of main on/off for channels 1-10 for the current faction on the current realm

`/aka main <name>` - Sets the name of your main for the current faction on the current realm

`/aka conf` - Displays the current configuration of AKA for the current faction on the current realm

`/aka on` - Enable the AKA AddOn for the current realm

`/aka off` - Disable the AKA AddOn for the current realm

`/aka reset` - Reset the current faction on the current realm to the default values


## Acknowledgements

AKA is inspired by the Identity AddOn that has filled a similar role over the years. Identity and its successor Identity-2 have been lovingly shepherded by Ferusnox and UltraNurd, and we thank them for their years of service to the WoW community.


## License

AKA is licensed under the Creative Commons Attribution 3.0 Unported (CC BY 3.0) license. For details, see LICENSE.txt located alongside this file. See also [http://creativecommons.org/licenses/by/3.0/legalcode] for the same content online.


CHANGELOG.txt
CHANGELOG for AKA

AKA 0.7.8
  * Updated for 8.3.0 compatibility

-- Barry Hawkins Tue, 12 May 2020 11:59:41 -0700

AKA 0.7.7
  * Updated for 8.2.5 compatibility

-- Barry Hawkins Sat, 02 Nov 2019 21:21:49 -0700

AKA 0.7.6
  * Updated for 8.1.0 compatibility

-- Barry Hawkins Mon, 25 Feb 2019 10:58:13 -0800

AKA 0.7.5
  * Updated for 8.0.1 compatibility

-- Barry Hawkins Wed, 18 Jul 2018 08:20:33 -0700

AKA 0.7.4
  * Updated for 7.3 compatibility

-- Barry Hawkins Sun, 10 Sep 2017 22:33:40 -0700

AKA 0.7.3
  * Updated for 7.2 compatibility

-- Barry Hawkins Sat, 15 Apr 2017 10:19:41 -0700

AKA 0.7.2
  * Updated for 7.1 compatibility

-- Barry Hawkins Thu, 27 Oct 2016 00:25:58 -0700

AKA 0.7.1
  * Correction to specified Interface version

-- Barry Hawkins Tue, 19 Jul 2016 22:43:54 -0700

AKA 0.7
  * Updated for 7.0.3 compatibility

-- Barry Hawkins Tue, 19 Jul 2016 08:02:07 -0700

AKA 0.6
  * Updated for 6.2.4 compatibility

-- Barry Hawkins Sun, 17 Jul 2016 23:47:45 -0700

AKA 0.5
  * Indicate when player is currently playing their main character

-- Barry Hawkins Sat, 02 Nov 2013 09:58:55 -0700

AKA 0.4
  * Addition of configuration display command
  * Enhanced load message; now displays whether AKA is enabled

-- Barry Hawkins Wed, 16 Oct 2013 23:02:05 -0700

AKA 0.3
  * Version bump for WoW 5.4 release

-- Barry Hawkins Tue, 10 Sep 2013 06:53:48 -0700

AKA 0.2
  * Addition of support for channels 1-10
  * Removal of newline character from help message; those are not allowed

-- Barry Hawkins Thu, 29 Aug 2013 13:07:55 -0700

AKA 0.1
  * Initial beta release
  * Basic support for adding name of main
  * Basic support for enabling guild and whisper channels
  * Rename of project to AKA

-- Barry Hawkins Mon, 26 Aug 2013 13:42:11 -0700


src/AKA.lua
-- AKA.lua

function AKA_InitializeChannelDefinitions()
    local channels = {}
    channels.whisper = "WHISPER"
    channels.party = "PARTY"
    channels.guild = "GUILD"
    channels.officer = "OFFICER"
    channels.raid = "RAID"
    channels.battleground = "BATTLEGROUND"
    channels.channel01 = 1
    channels.channel02 = 2
    channels.channel03 = 3
    channels.channel04 = 4
    channels.channel05 = 5
    channels.channel06 = 6
    channels.channel07 = 7
    channels.channel08 = 8
    channels.channel09 = 9
    channels.channel10 = 10
    return channels
end

local AKA_Version = "0.7.8"
local AKA_SystemSendChatMessage
local AKA_ChannelDefinitions = AKA_InitializeChannelDefinitions()
local AKA_CharacterName
local AKA_CharacterFaction
local AKA_LocalizedCharacterFaction
local AKA_CurrentRealm
local AKA_RealmFactionKey

function AKA_OnLoad(frame)
    frame:RegisterEvent("VARIABLES_LOADED")
    frame:RegisterEvent("PLAYER_ENTERING_WORLD")

    SlashCmdList["AKA"] = AKA_Controller
    SLASH_AKA1 = "/aka"
end

function AKA_OnEvent(frame, event)
    if (event == "VARIABLES_LOADED") then
        if (not AKASettings) then
            AKA_InitializeAKASettings()
        end

        AKA_SystemSendChatMessage = SendChatMessage
        SendChatMessage = AKA_DecorateSendChatMessage

        DEFAULT_CHAT_FRAME:AddMessage(string.format(AKA_VERSION_LOADED, AKA_Version, AKA_TrueFalseAsOnOff(AKASettings.enabled)), 0.4, 0.4, 1.0)
    end

    if (event == "PLAYER_ENTERING_WORLD") then
        AKA_CurrentRealm = AKA_GetCurrentRealm()
        AKA_CharacterFaction = AKA_GetCurrentFaction()
        AKA_CharacterName = AKA_GetCurrentCharacterName()
        AKA_RealmFactionKey = string.format("%s %s", AKA_CurrentRealm, AKA_CharacterFaction)

        if (not AKASettings.realms[AKA_RealmFactionKey]) then
            AKASettings.realms[AKA_RealmFactionKey] = AKA_RealmSettingsFactory()
        end
    end
end

function AKA_InitializeAKASettings()
    AKASettings = {}
    AKASettings.version = AKA_Version
    AKASettings.enabled = true
    AKASettings.format = "(%s)"
    AKASettings.realms = {}
end

function AKA_RealmSettingsFactory()
    local realmSettings = {}
    realmSettings.aliases = {}
    realmSettings.aliases.main = ""
    realmSettings.channels = {}
    realmSettings.channels.whisper = false
    realmSettings.channels.party = false
    realmSettings.channels.guild = false
    realmSettings.channels.officer = false
    realmSettings.channels.raid = false
    realmSettings.channels.battleground = false
    realmSettings.channels.channel01 = false
    realmSettings.channels.channel02 = false
    realmSettings.channels.channel03 = false
    realmSettings.channels.channel04 = false
    realmSettings.channels.channel05 = false
    realmSettings.channels.channel06 = false
    realmSettings.channels.channel07 = false
    realmSettings.channels.channel08 = false
    realmSettings.channels.channel09 = false
    realmSettings.channels.channel10 = false
    return realmSettings
end

function AKA_DecorateSendChatMessage(msg, system, language, channel)
    if AKASettings.enabled then
        if AKA_ChannelIsEnabled(system, channel) then
            msg = AKA_AddPrefix(msg)
        end
    end
    AKA_SystemSendChatMessage(msg, system, language, channel)
end

function AKA_AddPrefix(msg)
    if AKASettings.realms[AKA_RealmFactionKey].aliases.main == AKA_CharacterName then
        return string.format("(%s) %s", AKA_DICT_MAIN, msg)
    elseif AKASettings.realms[AKA_RealmFactionKey].aliases.main == "" then
        DEFAULT_CHAT_FRAME:AddMessage(AKA_ENABLED_MAIN_BLANK, 0.4, 0.4, 1.0)
        return msg
    else
        local prefix = string.format(AKASettings.format, AKASettings.realms[AKA_RealmFactionKey].aliases.main)
        return string.format("%s %s", prefix, msg)
    end
end

function AKA_ChannelIsEnabled(system, channel)
    if system == AKA_ChannelDefinitions.whisper then
        return AKASettings.realms[AKA_RealmFactionKey].channels.whisper
    elseif system == AKA_ChannelDefinitions.party then
        return AKASettings.realms[AKA_RealmFactionKey].channels.party
    elseif system == AKA_ChannelDefinitions.guild then
        return AKASettings.realms[AKA_RealmFactionKey].channels.guild
    elseif system == AKA_ChannelDefinitions.officer then
        return AKASettings.realms[AKA_RealmFactionKey].channels.officer
    elseif system == AKA_ChannelDefinitions.battleground then
        return AKASettings.realms[AKA_RealmFactionKey].channels.battleground
    elseif system == "CHANNEL" then
        if channel == AKA_ChannelDefinitions.channel01 then
            return AKASettings.realms[AKA_RealmFactionKey].channels.channel01
        elseif channel == AKA_ChannelDefinitions.channel02 then
            return AKASettings.realms[AKA_RealmFactionKey].channels.channel02
        elseif channel == AKA_ChannelDefinitions.channel03 then
            return AKASettings.realms[AKA_RealmFactionKey].channels.channel03
        elseif channel == AKA_ChannelDefinitions.channel04 then
            return AKASettings.realms[AKA_RealmFactionKey].channels.channel04
        elseif channel == AKA_ChannelDefinitions.channel05 then
            return AKASettings.realms[AKA_RealmFactionKey].channels.channel05
        elseif channel == AKA_ChannelDefinitions.channel06 then
            return AKASettings.realms[AKA_RealmFactionKey].channels.channel06
        elseif channel == AKA_ChannelDefinitions.channel07 then
            return AKASettings.realms[AKA_RealmFactionKey].channels.channel07
        elseif channel == AKA_ChannelDefinitions.channel08 then
            return AKASettings.realms[AKA_RealmFactionKey].channels.channel08
        elseif channel == AKA_ChannelDefinitions.channel09 then
            return AKASettings.realms[AKA_RealmFactionKey].channels.channel09
        elseif channel == AKA_ChannelDefinitions.channel10 then
            return AKASettings.realms[AKA_RealmFactionKey].channels.channel10
        end
    end
    
    return false
end

function AKA_Controller(msg)
    local command, parameters = AKA_ParseCommand(msg)

    if command == "on" then
        AKA_Enable(true)
    elseif command == "off" then
        AKA_Enable(false)
    elseif command == "" then
        AKA_DisplayHelp()
    elseif command == "help" then
        AKA_DisplayHelp()
    elseif command == "conf" then
        AKA_DisplayConfiguration()
    elseif command == "main" then
        AKA_SetMain(parameters)
    elseif command == "guild" then
        AKA_ToggleGuildChannel()
    elseif command == "whisper" then
        AKA_ToggleWhisperChannel()
    elseif (tonumber(command) and ((tonumber(command) >= 1) and (tonumber(command) <=10))) then
        AKA_ToggleNumericChannel(command)
    elseif command == "reset" then
        AKA_ResetDefaults()
    else
        AKA_CommandNotRecognized(command)
    end
end

function AKA_ParseCommand(msg)
    if msg then
        local s, e, command = string.find(msg, "(%S+)")
        if s then
            return command, string.sub(msg, e + 2)
        else
            return ""
        end
    end
end

function AKA_Enable(enable)
    if enable then
        DEFAULT_CHAT_FRAME:AddMessage(AKA_ENABLED, 0.4, 0.4, 1.0)
    else
        DEFAULT_CHAT_FRAME:AddMessage(AKA_DISABLED, 0.4, 0.4, 1.0)
    end

    AKASettings.enabled = enable
end

function AKA_SetMain(name)
    local oldName = AKASettings.realms[AKA_RealmFactionKey].aliases.main
    AKASettings.realms[AKA_RealmFactionKey].aliases.main = name
    DEFAULT_CHAT_FRAME:AddMessage(string.format(AKA_MAIN_SET_VALUE, oldName, name), 0.4, 0.4, 1.0)
end

function AKA_ToggleGuildChannel()
    if AKASettings.realms[AKA_RealmFactionKey].channels.guild == true then
        AKASettings.realms[AKA_RealmFactionKey].channels.guild = false
        DEFAULT_CHAT_FRAME:AddMessage(AKA_DISABLED_FOR_GUILD, 0.4, 0.4, 1.0)
    else
        AKASettings.realms[AKA_RealmFactionKey].channels.guild = true
        DEFAULT_CHAT_FRAME:AddMessage(AKA_ENABLED_FOR_GUILD, 0.4, 0.4, 1.0)
    end
end

function AKA_ToggleWhisperChannel()
    if AKASettings.realms[AKA_RealmFactionKey].channels.whisper == true then
        AKASettings.realms[AKA_RealmFactionKey].channels.whisper = false
        DEFAULT_CHAT_FRAME:AddMessage(AKA_DISABLED_FOR_WHISPER, 0.4, 0.4, 1.0)
    else
        AKASettings.realms[AKA_RealmFactionKey].channels.whisper = true
        DEFAULT_CHAT_FRAME:AddMessage(AKA_ENABLED_FOR_WHISPER, 0.4, 0.4, 1.0)
    end
end

function AKA_ToggleNumericChannel(channel)
    local channel_name = AKA_FormatNumericChannelName(channel)

    if AKASettings.realms[AKA_RealmFactionKey].channels[channel_name] == true then
        AKASettings.realms[AKA_RealmFactionKey].channels[channel_name] = false
        DEFAULT_CHAT_FRAME:AddMessage(string.format(AKA_DISABLED_FOR_CHANNEL, channel), 0.4, 0.4, 1.0)
    else
        AKASettings.realms[AKA_RealmFactionKey].channels[channel_name] = true
        DEFAULT_CHAT_FRAME:AddMessage(string.format(AKA_ENABLED_FOR_CHANNEL, channel), 0.4, 0.4, 1.0)
    end
end

function AKA_ResetDefaults()
    AKA_InitializeAKASettings()
    AKASettings.realms[AKA_RealmFactionKey] = AKA_RealmSettingsFactory()
    DEFAULT_CHAT_FRAME:AddMessage(AKA_RESET_TO_DEFAULT, 0.4, 0.4, 1.0)
end

function AKA_DisplayHelp()
    DEFAULT_CHAT_FRAME:AddMessage(AKA_HELP_PREAMBLE, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(AKA_HELP_NOARG, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(AKA_HELP_HELP, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(AKA_HELP_GUILD, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(AKA_HELP_WHISPER, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(AKA_HELP_CHANNEL, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(AKA_HELP_MAIN, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(AKA_HELP_CONF, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(AKA_HELP_ON, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(AKA_HELP_OFF, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(AKA_HELP_RESET, 0.4, 0.4, 1.0)
end

function AKA_DisplayConfiguration()
    local config = {}
    current_realm = AKASettings.realms[AKA_RealmFactionKey]

    config[#config + 1] = string.format(AKA_CONF_REALM_PREAMBLE, AKA_CurrentRealm, AKA_CharacterFaction)
    config[#config + 1] = string.format(AKA_CONF_LABEL_MAIN, current_realm.aliases.main)
    config[#config + 1] = string.format(AKA_CONF_LABEL_GUILD, AKA_TrueFalseAsOnOff(current_realm.channels.guild))
    config[#config + 1] = string.format(AKA_CONF_LABEL_PARTY, AKA_TrueFalseAsOnOff(current_realm.channels.party))
    config[#config + 1] = string.format(AKA_CONF_LABEL_WHISPER, AKA_TrueFalseAsOnOff(current_realm.channels.whisper))
    for i = 1, 10 do
        local channel_number = tostring(i)
        local channel_name = AKA_FormatNumericChannelName(tostring(i))
        config[#config + 1] = string.format(AKA_CONF_LABEL_CHANNEL, channel_number, AKA_TrueFalseAsOnOff(current_realm.channels[channel_name]))
    end
    config[#config + 1] = AKA_CONF_GLOBAL_PREAMBLE
    config[#config + 1] = string.format(AKA_CONF_LABEL_ENABLED, AKA_TrueFalseAsOnOff(AKASettings.enabled))

    DEFAULT_CHAT_FRAME:AddMessage(table.concat(config), 0.4, 0.4, 1.0)
end

function AKA_GetCurrentCharacterName()
    return UnitName("player")
end

function AKA_GetCurrentFaction()
    return UnitFactionGroup("player")
end

function AKA_GetCurrentRealm()
    return GetRealmName()
end

function AKA_CommandNotRecognized(command)
    DEFAULT_CHAT_FRAME:AddMessage(string.format(AKA_COMMAND_NOT_RECOGNIZED,command), 0.4, 0.4, 1.0)
end

function AKA_FormatNumericChannelName(channel)
    if #channel == 1 then
        return string.format("channel0%s", channel)
    else
        return string.format("channel%s", channel)
    end
end

function AKA_TrueFalseAsOnOff(value)
    if value == true then
        return "on"
    else
        return "off"
    end
end


dev_deploy.sh
#!/usr/bin/env sh

export WARCRAFT_MAC_ADDON_HOME="/Applications/World of Warcraft/_retail_/Interface/Addons"
export WARCRAFT_WIN_ADDON_HOME="/mnt/e/World of Warcraft/_retail_/Interface/addons"
export WARCRAFT_ADDON_HOME=""
export AKA_INSTALL_DIRECTORY=""
export AKA_STAGING_DIRECTORY="AKA"
export AKA_DEV_ARCHIVE_NAME="AKA-SNAPSHOT.zip"

if [ -d "${WARCRAFT_MAC_ADDON_HOME}" ]; then
  echo "Mac addon installation detected"
  WARCRAFT_ADDON_HOME="${WARCRAFT_MAC_ADDON_HOME}"
  echo "\n"
fi

if [ -d "${WARCRAFT_WIN_ADDON_HOME}" ]; then
  echo "Windows addon installation detected"
  WARCRAFT_ADDON_HOME="${WARCRAFT_WIN_ADDON_HOME}"
  echo "\n"
fi

AKA_INSTALL_DIRECTORY="${WARCRAFT_ADDON_HOME}/AKA"

if [ -d "${AKA_INSTALL_DIRECTORY}" ]; then
  echo "'${AKA_INSTALL_DIRECTORY}' exists, removing for fresh deploy"
  rm -rfv "${AKA_INSTALL_DIRECTORY}"
  echo "\n"
fi

if [ -d "${AKA_STAGING_DIRECTORY}" ]; then
	echo "'${AKA_STAGING_DIRECTORY}' exists; left over from stale build?"
	rm -rfv "${AKA_STAGING_DIRECTORY}"
	echo "\n"
fi

echo "Creating staging directory '${AKA_STAGING_DIRECTORY}'"
mkdir -v "${AKA_STAGING_DIRECTORY}"
echo "\n"

if [ -d "${AKA_STAGING_DIRECTORY}" ]; then
  echo "Installing snapshot into staging directory '${AKA_STAGING_DIRECTORY}':"
  cp -v src/AKA.* "${AKA_STAGING_DIRECTORY}"
  cp -v src/localization.* "${AKA_STAGING_DIRECTORY}"
  cp -v *.txt "${AKA_STAGING_DIRECTORY}"
  echo "\n"

  echo "Installing snapshot into installation directory '${AKA_INSTALL_DIRECTORY}':"
  cp -Rv ${AKA_STAGING_DIRECTORY} "${WARCRAFT_ADDON_HOME}"
  echo "\n"
else
  echo "'${AKA_STAGING_DIRECTORY}' doesn't seem to exist?"
  exit -1
fi

echo "Creating AddOn archive of snapshot"
zip -r ${AKA_DEV_ARCHIVE_NAME} ${AKA_STAGING_DIRECTORY}
echo "\n"

if [ -d "${AKA_STAGING_DIRECTORY}" ]; then
  echo "Cleaning up staging directory"
  rm -rfv ${AKA_STAGING_DIRECTORY}
  echo "\n"
fi

 echo "Deploy completed at `date +"%a, %d %b %Y %H:%M:%S %z"`."


src/localization.en.lua

AKA_VERSION_LOADED = "AKA %s loaded; currently it is turned %s."
AKA_ENABLED_MAIN_BLANK = "AKA enabled for this channel but main is blank"
AKA_ENABLED = "AKA enabled"
AKA_DISABLED = "AKA disabled"
AKA_DISABLED_FOR_GUILD = "AKA disabled for guild chat"
AKA_ENABLED_FOR_GUILD = "AKA enabled for guild chat"
AKA_DISABLED_FOR_WHISPER = "AKA disabled for whispers"
AKA_ENABLED_FOR_WHISPER = "AKA enabled for whispers"
AKA_DISABLED_FOR_CHANNEL = "AKA disabled for chat channel %s"
AKA_ENABLED_FOR_CHANNEL = "AKA enabled for chat channel %s"
AKA_RESET_TO_DEFAULT = "AKA has been reset to default settings for all realms"
AKA_COMMAND_NOT_RECOGNIZED = "AKA command '%s' not recognized"
AKA_MAIN_SET_VALUE = "AKA value for main updated from '%s' to '%s'"

AKA_HELP_PREAMBLE = "This is the help for the AKA AddOn. See README for more information"
AKA_HELP_NOARG = "/aka - Displays basic help information"
AKA_HELP_HELP = "/aka help - Displays basic help information"
AKA_HELP_GUILD = "/aka guild - Toggle display of main on/off for guild chat for the current faction on the current realm"
AKA_HELP_WHISPER = "/aka whisper - Toggle display of main on/off for whispers for the current faction on the current realm"
AKA_HELP_CHANNEL = "/aka <number> - Toggle display of main on/off for channels 1-10 for the current faction on the current realm"
AKA_HELP_MAIN = "/aka main <name> - Sets the name of your main for the current faction on the current realm"
AKA_HELP_CONF = "/aka conf - Displays the configuration of AKA for the current faction on the current realm"
AKA_HELP_ON = "/aka on - Enable the AKA AddOn for the current realm"
AKA_HELP_OFF = "/aka off - Disable the AKA AddOn for the current realm"
AKA_HELP_RESET = "/aka reset - Reset the current faction on the current realm to the default values"

AKA_CONF_REALM_PREAMBLE = "AKA Configuration for %s, %s: "
AKA_CONF_LABEL_MAIN = "Main: '%s', "
AKA_CONF_LABEL_GUILD = "Guild:%s, "
AKA_CONF_LABEL_PARTY = "Party:%s, "
AKA_CONF_LABEL_WHISPER = "Whisper:%s, "
AKA_CONF_LABEL_CHANNEL = "%s:%s, "
AKA_CONF_GLOBAL_PREAMBLE = "AKA Global Configuration: "
AKA_CONF_LABEL_ENABLED = "Enabled:%s"

AKA_DICT_MAIN = "Main"


src/Moniker.lua
-- Moniker.lua

function Moniker_InitializeChannelDefinitions()
    local channels = {}
    channels.whisper = "WHISPER"
    channels.party = "PARTY"
    channels.guild = "GUILD"
    channels.officer = "OFFICER"
    channels.raid = "RAID"
    channels.battleground = "BATTLEGROUND"
    channels.channel01 = "01"
    channels.channel02 = "02"
    channels.channel03 = "03"
    channels.channel04 = "04"
    channels.channel05 = "05"
    channels.channel06 = "06"
    channels.channel07 = "07"
    channels.channel08 = "08"
    channels.channel09 = "09"
    channels.channel10 = "10"
    return channels
end

local Moniker_Version = "0.1"
local Moniker_SystemSendChatMessage
local Moniker_ChannelDefinitions = Moniker_InitializeChannelDefinitions()
local Moniker_CharacterName
local Moniker_CharacterFaction
local Moniker_LocalizedCharacterFaction
local Moniker_CurrentRealm
local Moniker_RealmFactionKey

function Moniker_OnLoad(frame)
    frame:RegisterEvent("VARIABLES_LOADED")
    frame:RegisterEvent("PLAYER_ENTERING_WORLD")

    SlashCmdList["MONIKER"] = Moniker_Controller
    SLASH_MONIKER1 = "/moniker"
    SLASH_MONIKER2 = "/mnk"
end

function Moniker_OnEvent(frame, event)
    if (event == "VARIABLES_LOADED") then
        if (not MonikerSettings) then
            Moniker_InitializeMonikerSettings()
        end

        Moniker_SystemSendChatMessage = SendChatMessage
        SendChatMessage = Moniker_DecorateSendChatMessage

        DEFAULT_CHAT_FRAME:AddMessage(string.format(MONIKER_VERSION_LOADED, Moniker_Version), 0.4, 0.4, 1.0)
    end

    if (event == "PLAYER_ENTERING_WORLD") then
        Moniker_CurrentRealm = Moniker_GetCurrentRealm()
        Moniker_CharacterFaction = Moniker_GetCurrentFaction()
        Moniker_CharacterName = Moniker_GetCurrentCharacterName()
        Moniker_RealmFactionKey = string.format("%s %s", Moniker_CurrentRealm, Moniker_CharacterFaction)

        if (not MonikerSettings.realms[Moniker_RealmFactionKey]) then
            MonikerSettings.realms[Moniker_RealmFactionKey] = Moniker_RealmSettingsFactory()
        end
    end
end

function Moniker_InitializeMonikerSettings()
    MonikerSettings = {}
    MonikerSettings.version = "0.1"
    MonikerSettings.enabled = true
    MonikerSettings.format = "(%s)"
    MonikerSettings.realms = {}
end

function Moniker_RealmSettingsFactory()
    local realmSettings = {}
    realmSettings.monikers = {}
    realmSettings.monikers.main = ""
    realmSettings.channels = {}
    realmSettings.channels.whisper = false
    realmSettings.channels.party = false
    realmSettings.channels.guild = false
    realmSettings.channels.officer = false
    realmSettings.channels.raid = false
    realmSettings.channels.battleground = false
    realmSettings.channels.channel01 = false
    realmSettings.channels.channel02 = false
    realmSettings.channels.channel03 = false
    realmSettings.channels.channel04 = false
    realmSettings.channels.channel05 = false
    realmSettings.channels.channel06 = false
    realmSettings.channels.channel07 = false
    realmSettings.channels.channel08 = false
    realmSettings.channels.channel09 = false
    realmSettings.channels.channel10 = false
    return realmSettings
end

function Moniker_DecorateSendChatMessage(msg, system, language, channel)
    if MonikerSettings.enabled then
        if Moniker_ChannelIsEnabled(system, channel) then
            msg = Moniker_AddPrefix(msg)
        end
    end
    Moniker_SystemSendChatMessage(msg, system, language, channel)
end

function Moniker_AddPrefix(msg)
    if MonikerSettings.realms[Moniker_RealmFactionKey].monikers.main == Moniker_CharacterName then
        return msg
    elseif MonikerSettings.realms[Moniker_RealmFactionKey].monikers.main == "" then
        DEFAULT_CHAT_FRAME:AddMessage(MONIKER_ENABLED_MAIN_BLANK, 0.4, 0.4, 1.0)
        return msg
    else
        local prefix = string.format(MonikerSettings.format, MonikerSettings.realms[Moniker_RealmFactionKey].monikers.main)
        return string.format("%s %s", prefix, msg)
    end
end

function Moniker_ChannelIsEnabled(system, channel)
    if system == Moniker_ChannelDefinitions.whisper then
        return MonikerSettings.realms[Moniker_RealmFactionKey].channels.whisper
    elseif system == Moniker_ChannelDefinitions.party then
        return MonikerSettings.realms[Moniker_RealmFactionKey].channels.party
    elseif system == Moniker_ChannelDefinitions.guild then
        return MonikerSettings.realms[Moniker_RealmFactionKey].channels.guild
    elseif system == Moniker_ChannelDefinitions.officer then
        return MonikerSettings.realms[Moniker_RealmFactionKey].channels.officer
    elseif system == Moniker_ChannelDefinitions.battleground then
        return MonikerSettings.realms[Moniker_RealmFactionKey].channels.battleground
    elseif system == "CHANNEL" then
        if channel == Moniker_ChannelDefinitions.channel01 then
            return MonikerSettings.realms[Moniker_RealmFactionKey].channels.channel01
        elseif channel == Moniker_ChannelDefinitions.channel02 then
            return MonikerSettings.realms[Moniker_RealmFactionKey].channels.channel02
        elseif channel == Moniker_ChannelDefinitions.channel03 then
            return MonikerSettings.realms[Moniker_RealmFactionKey].channels.channel03
        elseif channel == Moniker_ChannelDefinitions.channel04 then
            return MonikerSettings.realms[Moniker_RealmFactionKey].channels.channel04
        elseif channel == Moniker_ChannelDefinitions.channel05 then
            return MonikerSettings.realms[Moniker_RealmFactionKey].channels.channel05
        elseif channel == Moniker_ChannelDefinitions.channel06 then
            return MonikerSettings.realms[Moniker_RealmFactionKey].channels.channel06
        elseif channel == Moniker_ChannelDefinitions.channel07 then
            return MonikerSettings.realms[Moniker_RealmFactionKey].channels.channel07
        elseif channel == Moniker_ChannelDefinitions.channel08 then
            return MonikerSettings.realms[Moniker_RealmFactionKey].channels.channel08
        elseif channel == Moniker_ChannelDefinitions.channel09 then
            return MonikerSettings.realms[Moniker_RealmFactionKey].channels.channel09
        elseif channel == Moniker_ChannelDefinitions.channel10 then
            return MonikerSettings.realms[Moniker_RealmFactionKey].channels.channel10
        end
    end
    
    return false
end

function Moniker_Controller(msg)
    local command, parameters = Moniker_ParseCommand(msg)

    if command == "on" then
        Moniker_Enable(true)
    elseif command == "off" then
        Moniker_Enable(false)
    elseif command == "" then
        Moniker_DisplayHelp()
    elseif command == "help" then
        Moniker_DisplayHelp()
    elseif command == "main" then
        Moniker_SetMain(parameters)
    elseif command == "guild" then
        Moniker_ToggleGuildChannel()
    elseif command == "whisper" then
        Moniker_ToggleWhisperChannel()
    elseif command == "reset" then
        Moniker_ResetDefaults()
    else
        Moniker_CommandNotRecognized(command)
    end
end

function Moniker_ParseCommand(msg)
    if msg then
        local s, e, command = string.find(msg, "(%S+)")
        if s then
            return command, string.sub(msg, e + 2)
        else
            return ""
        end
    end
end

function Moniker_Enable(enable)
    if enable then
        DEFAULT_CHAT_FRAME:AddMessage(MONIKER_ENABLED, 0.4, 0.4, 1.0)
    else
        DEFAULT_CHAT_FRAME:AddMessage(MONIKER_DISABLED, 0.4, 0.4, 1.0)
    end

    MonikerSettings.enabled = enable
end

function Moniker_SetMain(name)
    local oldName = MonikerSettings.realms[Moniker_RealmFactionKey].monikers.main
    MonikerSettings.realms[Moniker_RealmFactionKey].monikers.main = name
    DEFAULT_CHAT_FRAME:AddMessage(string.format(MONIKER_MAIN_SET_VALUE, oldName, name), 0.4, 0.4, 1.0)
end

function Moniker_ToggleGuildChannel()
    if MonikerSettings.realms[Moniker_RealmFactionKey].channels.guild == true then
        MonikerSettings.realms[Moniker_RealmFactionKey].channels.guild = false
        DEFAULT_CHAT_FRAME:AddMessage(MONIKER_DISABLED_FOR_GUILD, 0.4, 0.4, 1.0)
    else
        MonikerSettings.realms[Moniker_RealmFactionKey].channels.guild = true
        DEFAULT_CHAT_FRAME:AddMessage(MONIKER_ENABLED_FOR_GUILD, 0.4, 0.4, 1.0)
    end
end

function Moniker_ToggleWhisperChannel()
    if MonikerSettings.realms[Moniker_RealmFactionKey].channels.whisper == true then
        MonikerSettings.realms[Moniker_RealmFactionKey].channels.whisper = false
        DEFAULT_CHAT_FRAME:AddMessage(MONIKER_DISABLED_FOR_WHISPER, 0.4, 0.4, 1.0)
    else
        MonikerSettings.realms[Moniker_RealmFactionKey].channels.whisper = true
        DEFAULT_CHAT_FRAME:AddMessage(MONIKER_ENABLED_FOR_WHISPER, 0.4, 0.4, 1.0)
    end
end

function Moniker_ResetDefaults()
    Moniker_InitializeMonikerSettings()
    MonikerSettings.realms[Moniker_RealmFactionKey] = Moniker_RealmSettingsFactory()
    DEFAULT_CHAT_FRAME:AddMessage(MONIKER_RESET_TO_DEFAULT, 0.4, 0.4, 1.0)
end

function Moniker_DisplayHelp()
    DEFAULT_CHAT_FRAME:AddMessage(MONIKER_HELP_PREAMBLE, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(MONIKER_HELP_NOARG, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(MONIKER_HELP_HELP, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(MONIKER_HELP_GUILD, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(MONIKER_HELP_WHISPER, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(MONIKER_HELP_MAIN, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(MONIKER_HELP_ON, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(MONIKER_HELP_OFF, 0.4, 0.4, 1.0)
    DEFAULT_CHAT_FRAME:AddMessage(MONIKER_HELP_RESET, 0.4, 0.4, 1.0)
end

function Moniker_GetCurrentCharacterName()
    return UnitName("player")
end

function Moniker_GetCurrentFaction()
    return UnitFactionGroup("player")
end

function Moniker_GetCurrentRealm()
    return GetRealmName()
end

function Moniker_CommandNotRecognized(command)
    DEFAULT_CHAT_FRAME:AddMessage(string.format(MONIKER_COMMAND_NOT_RECOGNIZED,command), 0.4, 0.4, 1.0)
end
